[Processing] Split with lines

bstroebl opened the pull request:
Split with lines has been revised again. The algorithm now also accepts multi geometries as input and always returns multi geometries as output: If the geometry of a feature is split the output is still one feature whose multi geometry is made up of the split parts of the input geometry.
Help and tests have been updated.

Commit:
Remove debugging messages

nyalldawson committed:
Add method to QgsGeometry to remove interior rings from a polygon

With an optional area threshold

nyalldawson committed:
[processing] Consolidate 'delete holes' algorithms

Copy min area parameter from 'Fill holes' algorithm to 'delete
holes' algorithm.

Also:

- make algorithm maintain z/m values
- make algorithm work with curved geometries
- add unit tests

nirvn committed:
[color ramp] implement invert() function

nyalldawson committed:
Rename minimum area paremeter, fix docs for QgsCurvePolygon::removeInteriorRing

nirvn committed:
[FEATURE] color ramp button widget

nirvn committed:
color ramp button widget for gradiant fill and categorized renderer

Commit:
[FEATURE] Return multi geometries

accept multi geometries as input

borysiasty committed:
[bugfix] Fix broken docks customization. Fixes #8055

Commit:
Adapt help file

elpaso committed:
Added histogram values to the bindings

rldhont committed:
[BUGFIX] Filter Legend By Map Content doesn't maintain point displacement legend

Fixed #11572
Needs to be backported and adapted to version 2.14

nyalldawson commented on python/plugins/processing/algs/qgis/SplitWithLines.py:
```diff
@@ -81,107 +81,111 @@ def processAlgorithm(self, progress):
+                        while len(inGeoms) > 0:
+                            inGeom = inGeoms.pop()
+
+                            if inGeom.geometry() == None: # this has been encountered and created a run-time error
```
It's not that important - but you could use `inGeom.isEmpty()` here to make this test a bit more self-explanatory.

bstroebl replied:
@nyalldawson I tried but with my test data I get
`AttributeError: 'NoneType' object has no attribute 'isEmpty'`
Therefore I leave the code as it is

nyalldawson replied:
Did you do inGeom.geometry().isEmpty() or inGeom.isEmpty() (should be the second)

bstroebl replied:
you're right, it was `inGeom.geometry().isEmpty()`, `inGeom.isEmpty()` works

nyalldawson commented on python/plugins/processing/algs/qgis/SplitWithLines.py:
```diff
@@ -81,107 +81,111 @@ def processAlgorithm(self, progress):
-                                      aGeom.asPolyline()[0] != aGeom.asPolyline()[1])
-                            # sometimes splitting results in lines of zero length
+                if QgsWkbTypes.geometryType( aGeom.wkbType() )  == QgsWkbTypes.LineGeometry:
+                    firstVertex = aGeom.vertexAt(0)
```
This could be simplified to `passed = aGeom.geometry().numPoints() > 2`. It also avoids the possibility that the third vertex is validly located at 0,0.

nyalldawson commented on python/plugins/processing/algs/qgis/SplitWithLines.py:
```diff
@@ -81,107 +81,111 @@ def processAlgorithm(self, progress):
-                        outFeat.setGeometry(aGeom)
-                        writer.addFeature(outFeat)
+                    if not passed:
+                        passed = secondVertex != QgsPoint(0, 0) # has 2 vertices
```
Same here - use geometry().numPoints() to avoid the (0,0) comparison

nyalldawson replied:
Actually.... this whole block is rather confusingly written. I'd suggest changing it to something like:

```
num_points = aGeom.geometry().numPoints()
if num_points > 2:
    passed = True
elif num_points == 2:
    passed = not aGeom.geometry().isClosed() # tests if vertex 0 = vertex 1
else:
    # not enough vertices to be a valid line
    passed = False
```

bstroebl replied:
@nyalldawson I was not happy with this block either but the numPoints method slipped my attention because I was looking in QGSGeometry api. Will try.

nyalldawson commented:
Great! Are you having trouble with the gml test files? (I often do). Here's some things which I've found helps:
- manually edit the gml and make sure each fid is unique. Then exclude then feature id column from the test comparison (use
```
compare:
  fields:
    fid: skip
```
) to do this
- delete the all the extra files except the gml, and then reopen it in qgis to force the gfs to be completely regenerated

nyalldawson committed:
[FEATURE][processing] Remove duplicate fill holes algorithm

The Fill_holes script is duplicated by the DeleteHoles algorithm,
and the DeleteHoles algorithm has many advantages:
- it's a full algorithm rather than a script
- uses qgis geometry classes and maintains curves/z/m
- has unit tests
- doesn't rely on non-standard external library (shapely)

(marked as feature for inclusion in release notes)

nyalldawson committed:
[FEATURE][processing] Remove duplicate create points along lines script

This functionality is duplicated by the PointsAlongGeometry algorithm

Keep the algorithm version, since it has unit tests and supports things
like progress reports.

(marked as feature for inclusion in release notes)

nirvn committed:
add test coverage for qgscolorramp's invert() function

nirvn committed:
[symbology] migrate graduated renderer color ramp widget (#3815)

Commit:
Fix unset variable error for CHECK_GRASS_EXTRA_INCLUDE_DIRS macro

Commit:
simplify code containing vertex checks

Commit:
Check for empty geometry instead of None

Commit:
[Test] Ignore fid in test output

Commit:
[Test] Recreate expected as multi geometries

bstroebl commented:
@nyalldawson I removed the fid-check (although fid was unique) and recreated gfs files to no avail

nyalldawson commented:
Squashed and pushed as https://github.com/qgis/QGIS/commit/b02c6a32069004b1a7703724a81777a7e2452c3b with a fix to the test data.

nyalldawson commented:
@bstroebl I had a look and I think you were using an outdated test result for the split lines with same layer test. Thanks for your work on this!

bstroebl commented:
@nyalldawson I was under the impression that I created the outputs anew after changing the the algorithm to multi output 
thank you and be prepared I will improve `eliminate sliver polygons` algorithm in the same way :-)
